.github/copilot-instructions.md

```diff
@@ -0,0 +1,219 @@
+- **package.json**: Standard Node.js package configuration
+
+### Core Components
+
+#### 1. Entity System (`lib/entities/`, `lib/converters/`)
+- **Entities**: ioBroker objects converted to Home Assistant format
+- **Converters**: Transform ioBroker object types to HA entity types:
+  - `switch.js` - Switch entities
+  - `light.js` - Light entities with dimming, color, etc.
+  - `sensor.js` - Sensor readings
+  - `binary_sensor.js` - On/off sensors
+  - `climate.js` - Thermostats and HVAC
+  - `cover.js` - Blinds, shutters, garage doors
+  - `media_player.js` - Media players
+  - `weather.js` - Weather stations
+  - `camera.js` - Camera/image entities
+  - `alarm_control_panel.js` - Security systems
+
+#### 2. Server System (`lib/server.js`)
+- Express.js-based web server
+- WebSocket communication for real-time updates
+- Authentication and session management
+- API endpoints for frontend communication
+
+#### 3. Modules (`lib/modules/`)
+- **autoEntities.js**: Automatic entity detection from ioBroker objects
+- **browser_mod.js**: Browser control functionality (popup, notification, etc.)
+- **dashboard.js**: Dashboard management
+- **entityRegistry.js**: Entity registration and ID management
+- **history.js**: Historical data handling
+- **statisticsRecorder.js**: Long-term statistics
+- **persistentNotifications.js**: System notifications
+- **todo.js**: Task/shopping list functionality
+
+#### 4. Frontend (`hass_frontend/`)
+- Modified Home Assistant frontend
+- Custom cards and UI components
+- Static assets and resources
+
+## File Structure
+
+```
+/
+├── main.js                    # Adapter entry point
+├── io-package.json           # Adapter configuration
+├── package.json              # Node.js dependencies
+├── lib/                      # Core adapter logic
+│   ├── server.js            # Main server implementation
+│   ├── entities/            # Entity management
+│   ├── converters/          # Object-to-entity converters
+│   ├── modules/             # Feature modules
+│   └── tools.js             # Utility functions
+├── admin/                   # Admin UI files
+│   ├── words.js            # Translations
+│   └── i18n/               # Localization files
+├── hass_frontend/          # Modified Home Assistant frontend
+├── docs/                   # Documentation (multi-language)
+├── test/                   # Test files
+└── gulpfile.js            # Build tasks
+```
+
+## Development Patterns & Conventions
+
+### Entity Conversion Pattern
+```javascript
+// Standard pattern for converter functions
+function processEntityType(id, obj, entity, objects, customsData) {
+    // 1. Validate object structure
+    // 2. Extract states and attributes
+    // 3. Map to Home Assistant entity format
+    // 4. Apply custom configurations
+    // 5. Return formatted entity
+}
+```
+
+### Module Pattern
+```javascript
+// Standard module pattern used throughout
+class ModuleName {
+    constructor(adapter) {
```

Comment:
there are more functions, that a module can have: `processMessage`, `onStateChange`, `onObjectChange`, `removeSubscription`, `augmentServices`, `processServiceCall`.

Better add them. All of the functions are optional. But if a module needs one of the functions, the name should exactly match one of the names, so it will be automatically called in server.js